File: src/main/java/platform/ecommerce/domain/product/ProductOption.java
package platform.ecommerce.domain.product;

import jakarta.persistence.*;
import lombok.*;
import platform.ecommerce.domain.common.BaseEntity;
import platform.ecommerce.exception.ErrorCode;
import platform.ecommerce.exception.InvalidStateException;

import java.math.BigDecimal;

/**
 * Product option entity.
 * Represents variations like color, size with individual stock.
 */
@Entity
@Table(name = "product_option")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class ProductOption extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = false)
    private Product product;

    @Enumerated(EnumType.STRING)
    @Column(name = "option_type", nullable = false, length = 20)
    private OptionType optionType;

    @Column(name = "option_value", nullable = false, length = 100)
    private String optionValue;

    @Column(name = "additional_price", nullable = false, precision = 12, scale = 2)
    private BigDecimal additionalPrice;

    @Column(nullable = false)
    private int stock;

    @Column(name = "display_order", nullable = false)
    private int displayOrder;

    @Builder
    public ProductOption(Product product, OptionType optionType, String optionValue,
                         BigDecimal additionalPrice, int stock, int displayOrder) {
        validateStock(stock);
        this.product = product;
        this.optionType = optionType;
        this.optionValue = optionValue;
        this.additionalPrice = additionalPrice != null ? additionalPrice : BigDecimal.ZERO;
        this.stock = stock;
        this.displayOrder = displayOrder;
    }

    /**
     * Calculate total price including additional price.
     */
    public BigDecimal calculatePrice(BigDecimal basePrice) {
        return basePrice.add(additionalPrice);
    }

    /**
     * Decrease stock for purchase.
     */
    public void decreaseStock(int quantity) {
        if (quantity > this.stock) {
            throw new InvalidStateException(ErrorCode.INSUFFICIENT_STOCK);
        }
        this.stock -= quantity;
    }

    /**
     * Increase stock for cancellation or restock.
     */
    public void increaseStock(int quantity) {
        this.stock += quantity;
    }

    /**
     * Update stock directly.
     */
    public void updateStock(int stock) {
        validateStock(stock);
        this.stock = stock;
    }

    /**
     * Check if option is in stock.
     */
    public boolean isInStock() {
        return this.stock > 0;
    }

    /**
     * Update option details.
     */
    public void update(String optionValue, BigDecimal additionalPrice, int displayOrder) {
        this.optionValue = optionValue;
        this.additionalPrice = additionalPrice != null ? additionalPrice : BigDecimal.ZERO;
        this.displayOrder = displayOrder;
    }

    private void validateStock(int stock) {
        if (stock < 0) {
            throw new IllegalArgumentException("Stock cannot be negative");
        }
    }
}

File: src/main/kotlin/platform/ecommerce/domain/ProductOption.kt
package platform.ecommerce.domain

import jakarta.persistence.*
import jakarta.persistence.GenerationType.IDENTITY

@Entity
@Table(name = "product_option")
class ProductOption(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = false)
    val product: Product,

    @Column(nullable = false, unique = true, length = 100)
    var sku: String,

    @Column(nullable = false, name = "option_name")
    var optionName: String,

    @Column(nullable = false, name = "stock_quantity")
    private var stockQuantity: Int
) : BaseEntity() {

    @Id
    @GeneratedValue(strategy = IDENTITY)
    var id: Long? = null
        protected set

    fun getStockQuantity(): Int = stockQuantity

    fun decreaseStock(quantity: Int) {
        require(quantity > 0) { "Quantity must be positive" }
        require(stockQuantity >= quantity) { "Insufficient stock: requested=$quantity, available=$stockQuantity" }

        stockQuantity -= quantity
        product.checkAndUpdateStatus()
    }

    fun increaseStock(quantity: Int) {
        require(quantity > 0) { "Quantity must be positive" }

        stockQuantity += quantity
        product.checkAndUpdateStatus()
    }

    fun isOrderable(): Boolean {
        return stockQuantity > 0
    }

    fun canOrder(quantity: Int): Boolean {
        return stockQuantity >= quantity
    }

    companion object {
        fun create(
            product: Product,
            sku: String,
            optionName: String,
            stockQuantity: Int
        ): ProductOption {
            require(sku.isNotBlank()) { "SKU cannot be blank" }
            require(optionName.isNotBlank()) { "Option name cannot be blank" }
            require(stockQuantity >= 0) { "Stock quantity must be non-negative" }

            return ProductOption(
                product = product,
                sku = sku,
                optionName = optionName,
                stockQuantity = stockQuantity
            )
        }
    }
}
